### PR TITLE
upgrade process check for customisation first

### PR DESCRIPTION
**Pull request by gymad**
## Description

Upgrade process check for custom folder first and alert user if there is any related  customisation
image

(ref: issue #3610 )

- Upgrade process also log the affected filenames
- This PR update the licence header in language file

## Motivation and Context

System doesn't shows any alert if upgrades having customisation but it could affect for the current behavior

## How To Test This

Make some customisation and upgrade the system as usually

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [x] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

**Comment by Mausino** [comment]
@samus-aran we will test it on upgrade on 7.1.. we add it  into system now and we will know more later.. sorry i have not so much time in this days

**Comment by chris001** [comment]
@gymad 
This is a good step in the right direction. 👍 
I notice this PR checks only for the existence of ONE TYPE of customization, custom overrides, which are upgrade-safe code files used in place of the core code files.
When we first raised this issue, it was for the purpose of detecting all types of customization to all code flles.
To this end, it's essential to check every core file for modifications.
How? Use the data in `files.md5`, and compare the expected `md5` value to the actual value returned by `md5sum ( file_get_contents ( $filename ) )`.

File: modules/UpgradeWizard/uw_utils.php [hunk 1]
```diff
@@ -781,7 +781,69 @@ function upgradeUWFiles($file) {
 		}
```
Comment (daniel-samson) [inline modules/UpgradeWizard/uw_utils.php]: Please update the license header  for modules/UpgradeWizard/uw_utils.php

File: modules/UpgradeWizard/uw_utils.php [hunk 2]
```diff
@@ -781,7 +781,69 @@ function upgradeUWFiles($file) {
...
 	}
 
-    upgradeUWFilesCopy($allFiles, $from_dir);
+	// check custom changes and alert the user before upgrade
+
+	if($filesInCustom = checkCustomOverrides($from_dir)) {
+		global $mod_strings;
+		$alertMessage = $mod_strings["LBL_UPGRD_CSTM_CHK"]; //
```
Comment (daniel-samson) [inline modules/UpgradeWizard/uw_utils.php]: Do we nee the empty comment and the end of this line?

File: modules/UpgradeWizard/uw_utils.php [hunk 3]
```diff
@@ -781,7 +781,69 @@ function upgradeUWFiles($file) {
...
+		echo "</ul></div>";
+	}
+
+	upgradeUWFilesCopy($allFiles, $from_dir);
+}
+
+/**
+ * find files in custom folder if it also in upgrade pack
+ *
+ * @param $fromDir uploaded temp directory
```
Comment (daniel-samson) [inline modules/UpgradeWizard/uw_utils.php]: what type is $fromDir?  Can you please put the type in?

Comment (gymad) [inline modules/UpgradeWizard/uw_utils.php]: it was not is the PR but thank you for your notice :)

**Comment by chris001** [comment]
One big problem, it's not checking the checksums of the core files.  It should because admins want to know which of their core files have been modified.  Often they'll want to backup those modified core files and manually reapply the changed lines of code to the new distributed core files.

**Comment by pgorod** [comment]
Well this is a good start, at least.

Some of what @chris001 is asking for can be found already done in `upload/upgrades/patch/scripts/upgrade_utils.php` (file only found on systems that have been upgraded, it's part of the upgrade scripts that aren't yet public on GitHub).

See the function `cleanUpCustom`, it seems to be checking MD5's.

I have a file in the same directory called `customCleanUpFiles.php` that includes a lot of content in this form:

```php
$files = array(
    array('custom/Extension/application/Ext/EntryPointRegistry/FP_events_EntryPoint.php',array('dd57a5da8d93a893a25e9d9132bcde35'),false),
    array('custom/Extension/application/Ext/EntryPointRegistry/image_field.php',array('6221736e5c8f8087c91f626836199932'),false),
    array('custom/Extension/application/Ext/Extensions/aow_workflow.php',array('e057780e97b74486a9832947b5f5e6ea'),false),
    array('custom/Extension/application/Ext/Include/Advanced OpenEvents.php',array('dc2ce384c35b8b8844e1095c6fb88e98'),false),

(...etc...)
```
This supplies the material for that function I mentioned to check each file, although this is used for another purpose.

**Comment by gymad** [comment]
@chris001 You'r right but I don't think the manually non-upgrade safe changes related to any customisation. If these admins change core files they have to manage they own changes. There are many of tools for it for e.g version controlling, this PR related to the custom changes only.

**Pull request merged**